A87-mtls-spiffe-support.md

```diff
@@ -0,0 +1,200 @@
+----
+* Author(s): @erm-g, @markdroth
+* Approver: @ejona86, @dfawley
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2024-11-01
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+We will add support for SPIFFE certificate verification in mTLS.  This
+will be supported both with TlsCredentials and via xDS.
+
+## Background
+
+[SPIFFE] is a standardized way of encoding workload identity in mTLS
+certificates.  Specifically, we want to support using the [SPIFFE bundle
+format] in place of a single CA certificate for peer certificate
+verification.  The bundle map format is defined in
+https://github.com/spiffe/spiffe/pull/304.
+
+### Related Proposals: 
+* [gRFC A29: xDS-Based mTLS Security for gRPC Clients and Servers][gRFC A29]
+
+[gRFC A29]: A29-xds-tls-security.md
+[SPIFFE]: https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE.md
+[SPIFFE bundle format]: https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md#4-spiffe-bundle-format
+
+## Proposal
+
+There are two main parts to this proposal, support for SPIFFE
+certificate verification in TlsCredentials, and configuring that
+functionality via xDS.  We will cover the xDS case first.
+
+### Configuring SPIFFE Verification via xDS
+
+Currently, certificate providers as defined in [gRFC A29] can return
+only a single CA certificate for use in peer certificate verification.  We
+will change the certificate provider API such that the provider may
+choose to return either CA certificates or a SPIFFE trust bundle
+map, although it must choose one -- it may not return both for the same
+certificate name.
+
+If the certificate provider returns CA certificates, then
+standard X509 certificate verification will be used, just as it is
+today.  However, if the certificate provider returns a SPIFFE trust
+bundle map, then SPIFFE verification will be performed, and non-SPIFFE
+peer certificates will be rejected.
+
+When performing SPIFFE verification, the following steps will be performed:
+
+1. Upon receiving a peer certificate, verify that it is a well-formed SPIFFE
+   leaf certificate.  In particular, it must have a single URI SAN containing
+   a well-formed SPIFFE ID.
+
+2. Use the trust domain in the peer certificate's SPIFFE ID to lookup
+   the SPIFFE trust bundle. If the trust domain is not contained in the
+   configured trust map, reject the certificate.
+
+3. Verify the peer certificate using the default security library using
+   the SPIFFE trust bundle certificates as roots.
+
+SAN matching will be performed as configured via xDS, just as it is with
+X509 verification.
+
+gRPC currently supports only one certificate provider implementation,
+which is the `file_watcher` provider.  We will add a new field to the
+configuration of this provider implementation configuration called
+`spiffe_trust_bundle_map_file`.  This field will specify the path to
```

Comment:
Can we shorten the name to `spiffe_trust_bundle_file` (drop the `map`)?

Reply:
OK I see it's being called a "bundle map" so I guess it's needed. Sorry for the spam.